src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs

```diff
@@ -350,12 +351,46 @@ private DocumentState SetParseOptions(ParseOptions options)
                 throw new InvalidOperationException();
             }
 
-            var newTreeSource = CreateLazyFullyParsedTree(
-                TextAndVersionSource,
-                Id.ProjectId,
-                GetSyntaxTreeFilePath(Attributes),
-                options,
-                _languageServices);
+            ValueSource<TreeAndVersion> newTreeSource;
+
+            // Optimization: if we are only changing preprocessor directives, and we've already parsed the existing tree and it didn't have
+            // any, we can avoid a reparse since the tree will be parsed the same.
+            if (onlyPreprocessorDirectiveChange &&
+                _treeSource.TryGetValue(out var existingTreeAndVersion) &&
+                existingTreeAndVersion.Tree.TryGetRoot(out var existingRoot) &&
+                !existingRoot.ContainsDirectives)
+            {
+                var treeFactory = _languageServices.GetRequiredService<ISyntaxTreeFactoryService>();
+                SyntaxTree newTree;
+
+                if (treeFactory.CanCreateRecoverableTree(existingRoot))
+                {
+                    newTree = treeFactory.CreateRecoverableTree(Attributes.Id.ProjectId, FilePath, options, TextAndVersionSource, existingTreeAndVersion.Tree.Encoding, existingRoot);
+
+                    // The tree may be a RecoverableSyntaxTree. In its initial state, the RecoverableSyntaxTree keeps a
+                    // strong reference to the root SyntaxNode, and only transitions to a weak reference backed by temporary
+                    // storage after the first time GetRoot (or GetRootAsync) is called. Since we know we are creating a
+                    // RecoverableSyntaxTree for the purpose of avoiding problematic memory overhead, we call GetRoot
+                    // immediately to force the object to weakly hold its data from the start.
+                    // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1307180
+                    newTree.GetRoot(CancellationToken.None);
```

Comment:
@sharwell Do you recall why we didn't just fix CreateRecoverableTree to do this directly?

Reply:
(this is copy/pasted in several points in this file)

Reply:
This ceased to exist with the support for recoverable trees.